packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts

```diff
@@ -37,6 +37,51 @@ function isPropertyDefinitionWithTypeAnnotation(
   );
 }
 
+/**
+ * Checks if a node belongs to:
+ * ```
+ * foo(() => 1)
+ * ```
+ */
+function isFunctionArgument(
+  parent: TSESTree.Node,
+  callee?: FunctionExpression,
+): parent is TSESTree.CallExpression {
+  return (
+    parent.type === AST_NODE_TYPES.CallExpression &&
+    // make sure this isn't an IIFE
+    parent.callee !== callee
+  );
+}
+
+/**
+ * Checks if a node is an expression in a JSX attribute assignment
+ * ```
+ * <Comp x={...} />
+ * ```
+ */
+function isJSXAttribute(
+  node: TSESTree.Node,
+): node is TSESTree.JSXExpressionContainer | TSESTree.JSXSpreadAttribute {
+  return (
+    node.type === AST_NODE_TYPES.JSXExpressionContainer ||
```

Comment:
This is technically the relevant RHS type of `JSXAttribute`, so this check could also recurse one level higher if that is preferred. This should be the only way to have a function supplied in JSX though.

Reply:
I think we'll want to check the parent because this is another valid location for a `JSXExpressionContainer`:

```ts
<>{() => {}}</>
```

Reply:
@bradzacher Ah, I totally forgot about the render-props pattern. Running this down actually led me across an existing issue in TS where fragment shorthands are not typed properly: https://github.com/microsoft/TypeScript/issues/50429

I think all `JSXExpressionContainer` instances should be fully bound by types if it weren't for that issue ([reference CodeSandbox with render props](https://codesandbox.io/s/wispy-frog-sywpdm?file=/src/App.tsx)):

1. Attribute assignment: original case; type can be reliably inferred from assignment
1. Render props of built-ins and `<Fragment />`: not allowed due to limited child types (`JSX.ElementType` or `JSX.Element | null`); all functions will be type errors here
1. Render props of `<>` fragment shorthand: all functions _should_ be type errors here 
1. Render props of components with custom function children: properly typed-constrained by explicit `children` prop, the same as if children was defined via attribute

I went around on this for a while, but eventually realized that while something like this is unsafe:
```tsx
<>{() => 'bug'}</>
```
Having a lint error that turns it into this this doesn't make things better as it's still not a type error:
```tsx
<>{(): string => 'bug'}</>
```
While this is a currently an issue with TypeScript, it doesn't seem like the lint rule can do anything better here in place of proper type-checking.

[Reference TS-ESLint playground](https://typescript-eslint.io/play/#ts=5.2.1-rc&showAST=es&fileType=.tsx&code=MYewdgzgLgBAHjAvDAFAKBjAPAPg57HAbxQEokdUiYBDCAEwDMAuGKAJwFcBTGAX1J8sAejwFsAMXY0A5gFtuYKMTIUqtBizZdeAocKmyFSsQSwBhEHIAOK8okopqdJqw49%2BgkZZumReUgBuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)